---
title: "Genomics QC Report"
author: "GDC Data Team"
date: today
format: 
  html:
    toc: true
    html-math-method: katex
    embed-resources: true
execute:
  echo: false
  warning: false
  error: false
  message: false
  toc: true
---


```{r setup, include=FALSE}
library(tidyverse)
source("distViz.R")
source("logReader.R")
```

The following is an automated report describing the quality control (QC) processing of your dataset.
We first provide summary of the number of SNPs and subjects kept or removed followed by a more 
detailed description of the QC steps taken. 

# General Summary 
The QC process outline below first process the data in aggregate removing SNPs and subjects with large amounts
of missingness. In addition, SNPs are filtered to remove those with low minor allele frequency (MAF), frequencies
are two out of equlibrium based on Hardy-Weinberg equilibrium (HWE). Then, for multiethnic samples, subject 
ancestries are predicted using principal components analysis (PCA) and k-means clustering. Finally, the data
is filtered again at the level of each ancestry before it is later rejoined into the full dataset. Table @ref(tab:summary)
summarizes the number of SNPs and subjects at each step of the QC process.
```{r tab:summary, echo = FALSE}
inputData <- extractLog("../QCtutorial/logs/wgas2.log")
outputData <- extractLog("../QCtutorial/logs/wgas2.log")
outputData["OutSNPs"] <- outputData["OutSNPs"] - 13000
outputData["OutSubjects"] <- outputData["OutSubjects"] - 2
```
|Processing step | Number of SNPs | Number of Subjects |
|:---------------|---------------:|-------------------:|
|Input           | `r format(inputData["InSNPs"], scientific=FALSE)` | `r format(inputData["InSubjects"], scientific=FALSE)` |
|Output          | `r format(outputData["OutSNPs"], scientific=FALSE)` | `r format(outputData["OutSubjects"], scientific=FALSE)` |
|%Change         | `r format((outputData["OutSNPs"] - inputData["InSNPs"])/inputData["InSNPs"]*100, scientific=FALSE)` | `r format(outputData["OutSubjects"] -(inputData["InSubjects"] )/inputData["InSubjects"]*100, scientific=FALSE)` |

## Data set description
Case/controls


Ethnicities


## Data Processing
```{dot}
digraph G {
  input [ label = "1000 Subjects \n 10000 SNPs" ];
  filter [ label = "987 Subjects \n 9982 SNPs"];

  filter1 [label = "Ancestry 1 \n 335 Subjects \n 9980 SNPs"];
  filter2 [label = "Ancestry 2 \n 138 Subjects \n 9870 SNPs"];
  filter3 [label = "Ancestry 3 \n 514 Subjects \n 9901 SNPs"];

  joinedData[ label = " \n 987 Subjects \n 9862 SNPs"];

  input -> filter [ label = "MAF > 0.05 \n SNP Missingness < 0.1 \n SNP missingness < 0.05"];

  filter -> filter1, filter2;
  filter -> filter3 [label = "Predict ancestry"];

  filter1, filter2 -> joinedData; 
  filter3 -> joinedData[ label = "Matched and joined data"];
}
```


# Detailed QC summary
## Per Subject summary
Subjects were filtered based on missingness and heterozygosity. An LD pruning step was also performed.
```{r}
data.frame("Subject" = 1:100,
          "Missingness" = rbeta(100, 1, 4)) %>%
  ggplot(aes(x = Missingness)) +
  geom_histogram() + 
  geom_vline(xintercept = 0.15, color = "red") + 
  xlab("Missingness per subject") +
  ggtitle("% SNPS missing per subject")
```

The heterozygosity F statistic, often denoted as FIS, is a measure used in statistical genetics to assess the departure from Hardy-Weinberg equilibrium (HWE) in a population. Hardy-Weinberg equilibrium is a theoretical concept in population genetics that describes the expected genotype frequencies in a population when certain conditions are met, such as no mutation, migration, selection, or genetic drift.
The heterozygosity F statistic is given by the following equation:
$$
F = \frac{1}{N}\sum_{i=1}^N \frac{O_i - E_i}{E_i}
$$
```{r}
data.frame("Subject" = 1:100,
          "Heterozygosity" = rcauchy(100, 0.1, 0.1)) %>%
  ggplot(aes(x = Heterozygosity)) +
  geom_histogram() + 
  geom_vline(xintercept = c(-2, 2), color = "red") + 
  xlab("Heterozygosity F statistic") +
  ggtitle("Heterozygous F statistic distribution")
```

## Per SNP summary
SNPS were filtered based on missingness and MAF as well as HWE. 
Hardy-Weinberg equilibrium (HWE), also known as the Hardy-Weinberg law or principle, is a fundamental concept in population genetics that describes the expected genotype frequencies in a population when certain conditions are met. These conditions include the absence of factors like mutation, migration, selection, genetic drift, and non-random mating. In essence, it serves as a null model for understanding how genetic variation is maintained in populations under specific assumptions.

The Hardy-Weinberg equilibrium is based on the following principles:
- Random Mating: Individuals in the population mate randomly with respect to their genotype.
- No Mutation: There are no new alleles introduced into the population due to genetic mutations.
- No Migration: There is no migration of individuals into or out of the population.
- No Selection: There is no differential survival or reproductive advantage for individuals with particular genotypes.
- Infinitely Large Population: The population is considered infinitely large, which means that genetic drift (random fluctuations in allele frequencies) is negligible.

The formula for Hardy-Weinberg equilibrium relates genotype frequencies to allele frequencies. If we have two alleles for a particular gene, denoted as "A" and "a," and the allele frequencies are p for "A" and q for "a," then the expected genotype frequencies in the population are given by:

    Genotype AA frequency, $f(AA) = p^2$ 
    Genotype Aa frequency $f(Aa) = 2pq$
    Genotype aa frequency $f(aa) = q^2$

Where:

    p is the frequency of the "A" allele.
    q is the frequency of the "a" allele.
    p^2 represents the expected frequency of homozygous "AA" individuals.
    2pq represents the expected frequency of heterozygous "Aa" individuals.
    q^2 represents the expected frequency of homozygous "aa" individuals.

These genotype frequencies are expected to remain constant over generations when the conditions of Hardy-Weinberg equilibrium are met. Deviations from these expected frequencies can provide insights into evolutionary processes, such as genetic drift, natural selection, and mating patterns within a population. In practice, deviations from Hardy-Weinberg equilibrium often indicate that one or more of the assumptions are being violated, leading to evolutionary changes in the population's genetic composition.
```{r}
data.frame("SNP" = 1:1000,
          "Missingness" = rbeta(1000, 1, 5)) %>%
  ggplot(aes(x = Missingness)) +
  geom_histogram() +
  geom_vline(xintercept = 0.15, color = "red") +
  xlab("Missingness per SNP") +
  ggtitle("% calls missing per SNP")


data.frame("SNP" = 1:1000,
           "MAF" = rbeta(1000, 1, 3)) %>%
  ggplot(aes(x = MAF)) +
  geom_histogram() +
  geom_vline(xintercept = 0.05, color = "red") +
  xlab("Minor allele frequency") +
  ggtitle("MAF distribution")

data.frame("SNP" = 1:1000,
          "HWE" = -rgamma(1000, 3, 2)) %>%
          ggplot(aes(x = HWE)) +
          geom_histogram() +
          geom_vline(xintercept = -0.05, color = "red") +
          xlab("log(HWE p-value)") +
          ggtitle("Hardy-Weinberg Equilibrium p-value distribution")


```


## Ancestry predictions
Ancestries are predicted after the initial round of QC steps. First the standardized genetic relatedness matrix (GRM)
is calcualted by standardizing the genotype matrix and then calculating the outter product of the resulting matrix.
The GRM is then used to calculate the first 10 principal components (PCs). The subjects are then projected onto the 
PC space. The subjects are then assigned ethnicities using k-means clustering based on either a standard genomic dataset 
(such as though genomes) are by the reported ancestries of the subject in the study. Figure @ref(fig:ancestryPredictions)
the projection of each subject onto the first few PCs. They are colored by their reported ancestries. 

The densities of ancestries based on the reference set are represent by the 10\% contour lines of the 2d density plots
with the observed populations presented as points over top of the densities.
```{r ancestryPredictions}
data.frame(Ancestry = rep(c("Ancestry 1", "Ancestry 2"), each = 50),
           PC1 = c(rnorm(50, 0, 1), rnorm(50, 9, 1)),
           PC2 = c(rnorm(50, 0, 1), rnorm(50, 9, 1))) %>%
  ggplot(aes(x = PC1, y = PC2, color = Ancestry)) +
  geom_density_2d(breaks = c(0.05)) + 
  geom_point()
```



## Per Ancestry summary
We also summarize the QC metrics per ancestry. Figure @ref(fig:ancestrySummary) shows the distribution of MAF per ancestry.
```{r ancestrySummary}
data.frame("Ancestry" = rep(c("1", "2", "3"), each = 50),
           "MAF" = c(rbeta(50, 3, 2), rbeta(50, 3, 3), rbeta(50, 4, 1))) %>%
  ggplot(aes(x = MAF)) +
  geom_histogram() + 
  geom_vline(xintercept = 0.05, color = "red") + 
  xlab("Minor allele frequency") +
  ggtitle("Minor allele frequency per ancestry") + 
  facet_wrap(~Ancestry, ncol = 2)

```





# Additional information
## Referencing this pipeline
???? GDC citation????

## Licence
???? Confirm we want MIT license ????

## Contact
???? Who to contact for questions ????

